### PR TITLE
ima_dm: add type checks and hints

### PR DESCRIPTION
**Pull request by THS-on**
This adds type hints to all of the functions in ima_dm. I'm currently waiting for https://github.com/keylime/keylime/pull/1180 to finalize this PR.

### CODE REVIEW

File: keylime/ima/ima_dm.py [hunk 1]
```diff
@@ -7,9 +7,10 @@
 import pickle
 import re
 import sys
+import typing
 from collections import ChainMap
 from itertools import chain
-from typing import Dict, List, Optional, Union
+from typing import Any, Callable, Dict, List, Optional, TypeAlias, TypedDict, Union
```
Comment (aplanas) [inline keylime/ima/ima_dm.py:13]:
```
from typing import Any, Callable, Dict, List, Optional, TypeAlias, TypedDict, Union
ImportError: cannot import name 'TypeAlias' from 'typing' (/usr/lib64/python3.9/typing.py)
```

`New in version 3.10.` according to doc

Comment (THS-on) [reply keylime/ima/ima_dm.py]: Yeah I missed that this feature is only supported in Python 3.10. Removed it for now.

**Comment by THS-on** [comment]
`TypedDict` is also a Python 3.8 feature. How well is the `typing-extensions` package supported in distributions? 
@aplanas do have any SLES version that uses python 3.7?

**Comment by aplanas** [comment]
> `TypedDict` is also a Python 3.8 feature. How well is the `typing-extensions` package supported in distributions? @aplanas do have any SLES version that uses python 3.7

Right. For the next version we will have 3.10, and `typing-extensions` is already present.

**Comment by THS-on** [comment]
@kkaarreell @aplanas @ansasaki  what option do you prefer:

* Keep the patch as it is and add `typing-extensions` as a dependency
* Ship a backport of the TypedDict in Keylime
* Use a generic Dict type annotation

**Comment by kkaarreell** [comment]
We are not shipping keylime for RHEL-8 so if installing python-typing-extensions would do the job then we can install it from EPEL8. You can test that by installing the package in an extra step after
https://github.com/keylime/keylime/blob/master/packit-ci.fmf#L67
We should probably also discuss if there is still a value in running tests on C8S.

**Comment by aplanas** [comment]
> If we want to support Python 3.6 still, probably yes otherwise not really.

Do we have a process to drop old Python versions? For example, release a Keylime 7 based on Python 3.10 (that has almost 4 years of live ahead[1])

With a process like that we can pin in the enterprise distro the last supported Keylime branch (6.5 in this case).

A process like that will open the question about how many Keylime branches are supported by the project, so when a security fix happens in Keylime N, it is also backported to the N-1, N-2 ... branches.

[1] https://endoflife.date/python

**Comment by THS-on** [comment]
> Do we have a process to drop old Python versions?

Not yet, I think it is probably a good idea to raise this in the Slack channel.

File: keylime/ima/ima_dm.py [hunk 2]
```diff
@@ -904,11 +959,24 @@ def mirror_mirror_device_data(children):
...
+        ("dm_table_load", LoadEvent),
+        ("dm_device_resume", ResumeEvent),
+        ("dm_device_remove", RemoveEvent),
+        ("dm_table_clear", ClearEvent),
+        ("dm_device_rename", RenameEvent),
+        ("dm_target_update", UpdateEvent),
+    ]:
+        raise TypeError(f"{event} was parsed as: {type(event)}")
+
+    assert isinstance(out, (LoadEvent, ResumeEvent, RemoveEvent, ClearEvent, RenameEvent, UpdateEvent))
```
Comment (stefanberger) [inline keylime/ima/ima_dm.py:981]: The above should have this already covered...

Comment (THS-on) [reply keylime/ima/ima_dm.py:981]: Yes, but unfortunately mypy does not find it. I could use cast, but I don't know if this is the better solution.

Comment (stefanberger) [reply keylime/ima/ima_dm.py:981]: Could you move the type checking code into a function of its own and also apply it to those checks in `validate()`? I think that function could be called from both places? As long as pyright/mypy is happy afterwards...

Comment (THS-on) [reply keylime/ima/ima_dm.py:981]: @stefanberger I'll try to find a way to make this more elegant. The issue I don't think there is an elegant way to associate input values with the output type. @aplanas do you have an idea to solve this?

Comment (aplanas) [reply keylime/ima/ima_dm.py:981]: No idea!

Only a suggestion: if we create a function `def _from_str_to_even(event: str) -> EventTypes:` could this help?

Comment (THS-on) [reply keylime/ima/ima_dm.py:981]: It is probably a good idea to create that function, but I don't think we can get rid of the asserts in the validate function...

**Comment by stefanberger** [comment]
I am also running into the missing `TypedDict` issue: pr #1254 

One way to work around this would be to try to import TypedDict and if it works we could use it, otherwise define the new datatype to be the fallbacks of Dict[Any, Any] or Dict[str, Any] or similar. Once we get rid of older versions of python we could get rid of the try ... except sequences then. Though having this with later versions of python that we work with would really help.

File: keylime/ima/ima_dm.py (hunk 2 above)
Comment (THS-on) [reply keylime/ima/ima_dm.py:983]: @aplanas what do you think of the idea of introducing functions of the style `parse_LoadEvent(data: str) -> LoadEvent:`? The advantage is that we have type safety, the disadvantage is that we introduce 6 new functions that basically are doing the same thing.

Comment (THS-on) [reply keylime/ima/ima_dm.py:983]: Ok, then we will keep the asserts for now. Maybe in the future we find a better way to solve this.

**Comment by THS-on** [comment]
@kkaarreell it currently point to my fork of the tests that installs the typing-extensions module on RHEL. So when this gets merged this needs to be updated.

**Comment by kkaarreell** [comment]
> @kkaarreell it currently point to my fork of the tests that installs the typing-extensions module on RHEL. So when this gets merged this needs to be updated.

python3-typing-extensions v3.7.4.3 are available in EPEL both for C8S and C9S. I think I would rather install it from EPEL than from pip. I will prepare PR to test it.

File: packit-ci.fmf [hunk 3]
```diff
@@ -15,7 +15,7 @@
 
   discover:
     how: fmf
-    url: https://github.com/RedHat-SP-Security/keylime-tests
+    url: https://github.com/THS-on/keylime-tests
```
Comment (mpeters) [inline packit-ci.fmf:18]: What needs to change in the RH-SP-S tests for this to get merged?

Comment (mpeters) [reply packit-ci.fmf:18]: nvm

**Comment by kkaarreell** [comment]
> > @kkaarreell it currently point to my fork of the tests that installs the typing-extensions module on RHEL. So when this gets merged this needs to be updated.
> 
> python3-typing-extensions v3.7.4.3 are available in EPEL both for C8S and C9S. I think I would rather install it from EPEL than from pip. I will prepare PR to test it.

In my PR https://github.com/RedHat-SP-Security/keylime-tests/pull/276 I can see the same failure on C8S as the one present here
```
Traceback (most recent call last):
  File "/usr/local/bin/keylime_tenant", line 11, in <module>
    load_entry_point('keylime==6.5.2', 'console_scripts', 'keylime_tenant')()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 476, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2700, in load_entry_point
    return ep.load()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2318, in load
    return self.resolve()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2324, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/local/lib/python3.6/site-packages/keylime-6.5.2-py3.6.egg/keylime/cmd/tenant.py", line 3, in <module>
    from keylime import keylime_logging, tenant
  File "/usr/local/lib/python3.6/site-packages/keylime-6.5.2-py3.6.egg/keylime/tenant.py", line 20, in <module>
    from keylime.cli import options, policies
  File "/usr/local/lib/python3.6/site-packages/keylime-6.5.2-py3.6.egg/keylime/cli/policies.py", line 4, in <module>
    from keylime.ima import file_signatures, ima
  File "/usr/local/lib/python3.6/site-packages/keylime-6.5.2-py3.6.egg/keylime/ima/ima.py", line 17, in <module>
    from keylime.ima import ast, file_signatures, ima_dm
  File "/usr/local/lib/python3.6/site-packages/keylime-6.5.2-py3.6.egg/keylime/ima/ima_dm.py", line 36, in <module>
    class DeepChainMap(ChainMap[Any, Any]):  # pylint: disable=unsubscriptable-object
TypeError: 'ABCMeta' object is not subscriptable
```

File: requirements.txt [hunk 4]
```diff
@@ -20,3 +20,5 @@ pyasn1-modules>=0.2.1 # BSD
 jinja2>=3.0.0 # BSD
 # Keylime requires also gpg (LGPLv2+).
 # Those bindings must match the local GPG version and therefore this package is not installed via PyPI.
+# The typing-extensions module is currently required for Python versions < 3.8
```
Comment (kkaarreell) [inline requirements.txt:23]: Does it mean it won't be needed for C9S with Python 3.9.14, just for C8S?

Comment (stefanberger) [reply requirements.txt:23]: It depends on what features one wants from it.Some of the newer features are only available in 3.10 I read.

File: requirements.txt [hunk 5]
```diff
@@ -20,3 +20,5 @@ pyasn1-modules>=0.2.1 # BSD
...
+typing-extensions>=3.7.4 #PSF-2.0
```
Comment (aplanas) [inline requirements.txt:24]: Based on https://pip.pypa.io/en/stable/reference/requirement-specifiers/#examples

```
typing-extensions>=3.7.4 ; python_version < '3.8' #PSF-2.0
```

File: requirements.txt (hunk 4 above)
Comment (THS-on) [reply requirements.txt:23]: We are currently using Literal and TypedDict. If we use something else we can just bump the requirement.

File: requirements.txt (hunk 5 above)
Comment (THS-on) [reply requirements.txt]: Done

**Comment by THS-on** [comment]
@kkaarreell this should be fixed now.

**Review by aplanas — APPROVED** [review]

**Comment by THS-on** [comment]
Hmm now it does not find the module. I have no idea why, because it should be installed.

**Comment by THS-on** [comment]
> Thank you @THS-on , tests are passing in my PR also for C8S.

Ok, I now dropped the commit that temporarily changed it to my branch.

**Comment by kkaarreell** [comment]
> > Thank you @THS-on , tests are passing in my PR also for C8S.
> 
> Ok, I now dropped the commit that temporarily changed it to my branch.

I have merged my PR so you can rerun Packit tests against the main branch.

**Review by mpeters — APPROVED** [review]

**Pull request merged**